cmd/juju/action/common.go

```diff
@@ -197,18 +212,33 @@ func (c *runCommandBase) waitForTasks(ctx *cmd.Context, tasks []enqueuedAction,
 		}
 		if err != nil {
 			if errors.IsTimeout(err) {
-				return c.handleTimeout(tasks, resultReceivers)
+				return -1, c.handleTimeout(tasks, resultReceivers)
 			}
-			return errors.Trace(err)
+			return -1, errors.Trace(err)
 		}
 
 		resultReceivers.Add(result.receiver)
-		d := formatActionResult(result.task, actionResult, c.utc)
-		d["id"] = result.task // Action ID is required in case we timed out.
-		info[result.receiverId()] = d
+		resultData, resultExitCode := formatActionResult(result.task, actionResult, c.utc)
+		resultData["id"] = result.task // Action ID is required in case we timed out.
+		info[result.receiverId()] = resultData
+
+		// If any of the actions have a error exit code, then inform the user
+		// that their exec failed.
+		if resultExitCode < 1 {
+			continue
+		}
+
+		// If the number of tasks is just 1, show the underlying exit code. We
+		// can't do this for multiple tasks as we don't know what the
+		// appropriate code to show (highest or lowest)?
+		if len(tasks) == 1 {
```

Comment:
I believe we should be explicit with our errors and indicate which tasks failed and with what exit code. It will make the error much more verbose but at least it tells us what failed so we can retry it.

Reply:
We don't know essentially and exit code is optional.

Reply:
Don't we know (per task) since we iterate the result list? Do you happen to know if the optional exit code is by design and if so what was the use case that had that as a requirement?

Reply:
In fact, the stderr has ALL this information:

```
juju exec -m controller --all 'cat /etc/systemd/system/juju-db.service'
Running operation 29 with 3 tasks
  - task 30 on machine-0
  - task 31 on machine-1
  - task 32 on machine-2

Waiting for task 30...
Waiting for task 31...
Waiting for task 32...
cat: /etc/systemd/system/juju-db.service: No such file or directory

ERROR tasks failed with exit code: 1
```